File: 02-authentication-protocols/examples/decode-jwt.js
#!/usr/bin/env node
/**
 * JWT Decoder and Validator (Node.js)
 * Decodes JWT tokens and validates claims.
 * Usage: node decode-jwt.js <jwt_token>
 */

function base64urlDecode(str) {
  str = str.replace(/-/g, '+').replace(/_/g, '/');
  while (str.length % 4) str += '=';
  return Buffer.from(str, 'base64').toString('utf8');
}

function decodeJWT(token) {
  const parts = token.trim().split('.');
  if (parts.length !== 3) {
    throw new Error(`Invalid JWT: expected 3 parts, got ${parts.length}`);
  }

  return {
    header: JSON.parse(base64urlDecode(parts[0])),
    payload: JSON.parse(base64urlDecode(parts[1])),
    signature: parts[2],
  };
}

function validateClaims(payload) {
  const issues = [];
  const now = Math.floor(Date.now() / 1000);

  // Check expiration
  if (payload.exp) {
    const expDate = new Date(payload.exp * 1000).toISOString();
    if (payload.exp < now) {
      issues.push(`EXPIRED: Token expired at ${expDate}`);
    } else {
      const remaining = payload.exp - now;
      const mins = Math.floor(remaining / 60);
      issues.push(`OK: Expires at ${expDate} (${mins} minutes remaining)`);
    }
  } else {
    issues.push("WARNING: No 'exp' claim - token never expires");
  }

  // Check not-before
  if (payload.nbf) {
    const nbfDate = new Date(payload.nbf * 1000).toISOString();
    if (payload.nbf > now) {
      issues.push(`NOT YET VALID: Valid from ${nbfDate}`);
    } else {
      issues.push(`OK: Valid since ${nbfDate}`);
    }
  }

  // Check issued-at
  if (payload.iat) {
    const iatDate = new Date(payload.iat * 1000).toISOString();
    if (payload.iat > now) {
      issues.push(`WARNING: Issued in the future: ${iatDate}`);
    } else {
      issues.push(`OK: Issued at ${iatDate}`);
    }
  }

  // Check required claims
  const recommended = ['sub', 'iss', 'aud'];
  for (const claim of recommended) {
    if (!(claim in payload)) {
      issues.push(`WARNING: Missing recommended claim '${claim}'`);
    }
  }

  // Check algorithm in header
  return issues;
}

function formatOutput(decoded) {
  console.log('='.repeat(60));
  console.log('JWT DECODER (Node.js)');
  console.log('='.repeat(60));

  console.log('\n--- HEADER ---');
  console.log(JSON.stringify(decoded.header, null, 2));

  console.log('\n--- PAYLOAD ---');
  console.log(JSON.stringify(decoded.payload, null, 2));

  console.log('\n--- SIGNATURE ---');
  const sig = decoded.signature;
  console.log(sig.length > 50 ? sig.substring(0, 50) + '...' : sig);

  console.log('\n--- VALIDATION ---');
  const issues = validateClaims(decoded.payload);
  issues.forEach(issue => console.log(`  ${issue}`));

  console.log('\n' + '='.repeat(60));
}

// Main
const token = process.argv[2];
if (!token) {
  console.log('Usage: node decode-jwt.js <jwt_token>');
  console.log('\nExample:');
  console.log('  node decode-jwt.js eyJhbGciOiJSUzI1NiIs...');
  process.exit(1);
}

try {
  const decoded = decodeJWT(token);
  formatOutput(decoded);
} catch (err) {
  console.error(`Error decoding JWT: ${err.message}`);
  process.exit(1);
}

File: 02-authentication-protocols/examples/decode-jwt.py
#!/usr/bin/env python3
"""
JWT Decoder and Validator
Decodes JWT tokens and validates claims without external dependencies.
Usage: python decode-jwt.py <jwt_token>
"""

import json
import base64
import sys
import hmac
import hashlib
from datetime import datetime, timezone


def base64url_decode(data: str) -> bytes:
    """Decode Base64URL-encoded string."""
    padding = 4 - len(data) % 4
    if padding != 4:
        data += "=" * padding
    return base64.urlsafe_b64decode(data)


def decode_jwt(token: str) -> dict:
    """Decode a JWT token without validation (for inspection)."""
    parts = token.strip().split(".")

    if len(parts) != 3:
        raise ValueError(f"Invalid JWT format: expected 3 parts, got {len(parts)}")

    header = json.loads(base64url_decode(parts[0]))
    payload = json.loads(base64url_decode(parts[1]))
    signature = parts[2]

    return {
        "header": header,
        "payload": payload,
        "signature": signature,
    }


def validate_claims(payload: dict) -> list:
    """Validate standard JWT claims and return any issues found."""
    issues = []
    now = datetime.now(timezone.utc).timestamp()

    # Check expiration
    if "exp" in payload:
        exp_time = datetime.fromtimestamp(payload["exp"], tz=timezone.utc)
        if payload["exp"] < now:
            issues.append(f"EXPIRED: Token expired at {exp_time.isoformat()}")
        else:
            issues.append(f"OK: Expires at {exp_time.isoformat()}")
    else:
        issues.append("WARNING: No 'exp' claim - token never expires")

    # Check not-before
    if "nbf" in payload:
        nbf_time = datetime.fromtimestamp(payload["nbf"], tz=timezone.utc)
        if payload["nbf"] > now:
            issues.append(f"NOT YET VALID: Valid from {nbf_time.isoformat()}")
        else:
            issues.append(f"OK: Valid since {nbf_time.isoformat()}")

    # Check issued-at
    if "iat" in payload:
        iat_time = datetime.fromtimestamp(payload["iat"], tz=timezone.utc)
        if payload["iat"] > now:
            issues.append(f"WARNING: Issued in the future: {iat_time.isoformat()}")
        else:
            issues.append(f"OK: Issued at {iat_time.isoformat()}")

    # Check required claims
    required = ["sub", "iss", "aud"]
    for claim in required:
        if claim not in payload:
            issues.append(f"WARNING: Missing recommended claim '{claim}'")

    return issues


def format_output(decoded: dict) -> str:
    """Format decoded JWT for display."""
    lines = []
    lines.append("=" * 60)
    lines.append("JWT DECODER")
    lines.append("=" * 60)

    lines.append("\n--- HEADER ---")
    lines.append(json.dumps(decoded["header"], indent=2))

    lines.append("\n--- PAYLOAD ---")
    lines.append(json.dumps(decoded["payload"], indent=2))

    lines.append("\n--- SIGNATURE ---")
    lines.append(decoded["signature"][:50] + "..." if len(decoded["signature"]) > 50 else decoded["signature"])

    lines.append("\n--- VALIDATION ---")
    for issue in validate_claims(decoded["payload"]):
        lines.append(f"  {issue}")

    lines.append("\n" + "=" * 60)
    return "\n".join(lines)


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python decode-jwt.py <jwt_token>")
        print("\nExample:")
        print("  python decode-jwt.py eyJhbGciOiJSUzI1NiIs...")
        print("  echo $TOKEN | python decode-jwt.py -")
        sys.exit(1)

    token = sys.argv[1]
    if token == "-":
        token = sys.stdin.read().strip()

    try:
        decoded = decode_jwt(token)
        print(format_output(decoded))
    except Exception as e:
        print(f"Error decoding JWT: {e}", file=sys.stderr)
        sys.exit(1)
